### PR TITLE
SPU: Embedded PRX/overlay image patching, Implement images dumping, Image search improvements

### PR DESCRIPTION
**Pull request by elad335**
* Search SPU images in PRX/overlay modules, allows to patch cellSpurs' SPU images.
* Remove stupid restriction in SPU image search, check if any address contains SPU image and not just aligned 128-byte addresses.
* Dump SPU images into actual files if "SPU Debug" option is enabled in config.
* Optimize SPU image search to be extremely fast.
* Search SPU images in all PS3 binary LOAD segements, not just the first.

### CODE REVIEW

**Comment by elad335** [comment]
SPU images will be dumped to "rpcs3/spu_progs/" with the filename as the image hash that is logged.

**Comment by elad335** [comment]
MGS4 images from sys_overlay are being able to be patched as well now:

image

**Comment by elad335** [comment]
Other firmware SPU images, e.g. audio/video decoding modules:

W {PPU[0x1000000] Thread (main_thread) [0x017609ac]} sys_prx: _sys_prx_load_module(path=“/dev_flash/sys/external/libm4aacdec.sprx”, flags=0x0, pOpt=*0x0)
W {PPU[0x1000000] Thread (main_thread) [0x017609ac]} ppu_loader: **** Loaded to 0x1aa0000...0x1b1ffff (size=0x80000)
W {PPU[0x1000000] Thread (main_thread) [0x017609ac]} ppu_loader: **** Loaded to 0x1b20000...0x1b284df (size=0x84e0)
W {PPU[0x1000000] Thread (main_thread) [0x017609ac]} ppu_loader: Library cellM4AacDec_Library (rtoc=0x1b30400):
S {PPU[0x1000000] Thread (main_thread) [0x017609ac]} ppu_loader: SPU executable hash: SPU-99081a775a96b1e6277ec02fb1f854e08272d7f0 (<- 0)
Segment: p_type=0x1, p_vaddr=0x3000, p_filesz=0x1e530, p_memsz=0x1e530, p_offset=0x100
Segment: p_type=0x1, p_vaddr=0x21580, p_filesz=0xa0, p_memsz=0x1920, p_offset=0x1e680
Segment: p_type=0x4, p_vaddr=0x0, p_filesz=0x34, p_memsz=0x34, p_offset=0x1e720
SPUNAME: 'm4aacdec_spu.elf'
S {PPU[0x1000000] Thread (main_thread) [0x017609ac]} ppu_loader: SPU executable hash: SPU-fafa42207668a5e71af1b1174de67ab9919dbd5d (<- 0)
Segment: p_type=0x1, p_vaddr=0x3000, p_filesz=0x198c0, p_memsz=0x198c0, p_offset=0x100
Segment: p_type=0x1, p_vaddr=0x1c900, p_filesz=0x250, p_memsz=0x1170, p_offset=0x19a00
Segment: p_type=0x4, p_vaddr=0x0, p_filesz=0x34, p_memsz=0x34, p_offset=0x19c50
SPUNAME: 'm4aacdec_spu_sbr.elf'
image

File: rpcs3/Emu/Cell/PPUModule.cpp [hunk 1]
```diff
@@ -694,6 +694,91 @@ static auto ppu_load_imports(std::vector<ppu_reloc>& relocs, ppu_linkage_info* l
...
+		for (u32 i = 0; i < sizeof(sha1_hash); i++)
+		{
+			constexpr auto pal = "0123456789abcdef";
+			hash[4 + i * 2] = pal[sha1_hash[i] >> 4];
+			hash[5 + i * 2] = pal[sha1_hash[i] & 15];
+		}
+
+		if (g_cfg.core.spu_debug)
+		{
+			fs::file dump_file(fs::get_cache_dir() + "/spu_progs/" + hash, fs::rewrite);
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/PPUModule.cpp:761]: perhaps, SPUNAME could be appended to hash somehow.

**Review by Nekotekina — CHANGES_REQUESTED** [review]

File: rpcs3/Emu/Cell/PPUModule.cpp (hunk 1 above)
Comment (elad335) [reply rpcs3/Emu/Cell/PPUModule.cpp:761]: Do we really want multiple files with the same data in such case? Theres also a case where there are multiple NOTE sections...
Im also not modifiying current hash for this as it breaks patches.

Comment (elad335) [reply rpcs3/Emu/Cell/PPUModule.cpp:761]: Actually line 740 already does this.

Comment (Nekotekina) [reply rpcs3/Emu/Cell/PPUModule.cpp:761]: You misunderstood, I meant appending it to filename while keeping the hash. Sorry for the confusion.

File: rpcs3/Emu/Cell/PPUModule.cpp [hunk 2]
```diff
@@ -694,6 +694,106 @@ static auto ppu_load_imports(std::vector<ppu_reloc>& relocs, ppu_linkage_info* l
...
+		for (u32 i = 0; i < sizeof(sha1_hash); i++)
+		{
+			constexpr auto pal = "0123456789abcdef";
+			hash[4 + i * 2] = pal[sha1_hash[i] >> 4];
+			hash[5 + i * 2] = pal[sha1_hash[i] & 15];
+		}
+
+		if (g_cfg.core.spu_debug)
+		{
+			for (auto& name : names) name = vfs::escape(name, true);
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/PPUModule.cpp:775]: Not very good idea to make one-line loop, otherwise it's ok.

**Pull request merged**